src/ui/ui_common.rs

```diff
@@ -1,120 +1,362 @@
+use cursive::event::{Event, EventResult, Key};
+use cursive::theme::{BaseColor, Color, ColorStyle, ColorType, Effect, PaletteColor, Style};
+use cursive::traits::*;
+use cursive::utils::span::SpannedString;
+use cursive::view::{Margins, Offset, Position};
+
+use cursive::views::{
+  Dialog, DialogFocus, Layer, OnEventView, PaddedView, SelectView, ShadowView, TextView,
+};
+use cursive::{Cursive, CursiveRunnable};
+use std::sync::{Arc, Mutex};
+use std::thread;
+
+struct ModelData {
+  credits: Store<Option<i64>>,
+  machines: Store<Option<DrinkList>>,
+  api: API,
 }
 
-pub fn launch() {
-  /* Setup ncurses. */
-  initscr();
-  raw();
+// This should really get cleaned up:
+type Model = Arc<Mutex<ModelData>>;
 
-  /* Allow for extended keyboard (like F1). */
-  keypad(stdscr(), true);
-  noecho();
+/// Entrypoint, CLI will call this when we start up!
+pub fn launch(api: API) -> Result<(), APIError> {
+  api.get_token()?;
+  let mut siv = cursive::default();
+  let model = Arc::new(Mutex::new(ModelData {
+    credits: Store::new(None),
+    machines: Store::new(None),
+    api,
+  }));
 
-  /* Invisible cursor. */
-  curs_set(CURSOR_VISIBILITY::CURSOR_INVISIBLE);
+  // Nice to have
+  siv.add_global_callback('q', |s| s.quit());
 
-  /* Colors. */
-  do_color();
-  attron(COLOR_PAIR(2));
+  csh_logo(&mut siv);
 
-  /* Update the screen. */
-  refresh();
-}
+  let padding = credit_count(Arc::clone(&model), &mut siv);
 
-pub fn end() {
-  endwin();
-}
+  machine_list(Arc::clone(&model), &mut siv, padding);
 
-pub fn create_win(y: i32, x: i32, height: i32, width: i32) -> WINDOW {
-  let win = newwin(height, width, y, x);
-  box_(win, 0, 0);
-  wrefresh(win);
-  win
-}
+  let status_handle = {
+    let model = Arc::clone(&model);
+    let cb_sink = siv.cb_sink().clone();
+    thread::spawn(move || {
+      let api = model.lock().unwrap().api.clone();
+      match api.get_status_for_machine(None) {
+        Ok(machine_list) => {
+          let model = Arc::clone(&model);
+          cb_sink
+            .send(Box::new(move |siv| {
+              model.lock().unwrap().machines.set(siv, Some(machine_list));
+            }))
+            .unwrap();
+          Ok(())
+        }
+        Err(err) => Err(err),
+      }
+    })
+  };
+
+  let credits_handle = {
+    let model = Arc::clone(&model);
+    let cb_sink = siv.cb_sink().clone();
+    thread::spawn(move || {
+      let api = model.lock().unwrap().api.clone();
+      match api.get_credits() {
+        Ok(credit_count) => {
+          let model = Arc::clone(&model);
+          cb_sink
+            .send(Box::new(move |siv| {
+              model.lock().unwrap().credits.set(siv, Some(credit_count));
+            }))
+            .unwrap();
+          Ok(())
+        }
+        Err(err) => Err(err),
+      }
+    })
+  };
 
-pub fn refresh_win(win: WINDOW) {
-  box_(win, 0, 0);
-  wrefresh(win);
+  siv.run();
+
+  status_handle.join().unwrap()?;
+  credits_handle.join().unwrap()?;
```

Comment:
Is there a specific reason this is multi-threaded? Does cursive want you to do this? Just seems overkill without knowing the details

Reply:
I want to run both requests in parallel